internal/gcsx/client_readers/gcs_reader.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/util"
+)
+
+// ReaderType represents different types of go-sdk gcs readers.
+type ReaderType int
+
+// ReaderType enum values.
+const (
+	MB = 1 << 20
+
+	// Min read size in bytes for random reads.
+	// We will not send a request to GCS for less than this many bytes (unless the
+	// end of the object comes first).
+	minReadSize = MB
+
+	// Minimum number of seeks before evaluating if the read pattern is random.
+	minSeeksForRandom = 2
+)
+
+// ReaderType enum values.
+const (
+	// RangeReaderType corresponds to NewReader method in bucket_handle.go
+	RangeReaderType ReaderType = iota
+
+	// MultiRangeReaderType corresponds to NewMultiRangeDownloader method in bucket_handle.go
+	MultiRangeReaderType
+)
+
+type GCSReader struct {
+	gcsx.Reader
+	object *gcs.MinObject
+	bucket gcs.Bucket
+
+	rangeReader *RangeReader
+	mrr         *MultiRangeReader
+
+	// ReadType of the reader. Will be sequential by default.
+	readType string
+
+	sequentialReadSizeMb int32
+	seeks                uint64
+	totalReadBytes       uint64
+}
+
+func NewGCSReader(obj *gcs.MinObject, bucket gcs.Bucket, metricHandle common.MetricHandle, mrdWrapper *gcsx.MultiRangeDownloaderWrapper, sequentialReadSizeMb int32) *GCSReader {
+	return &GCSReader{
+		object:               obj,
+		bucket:               bucket,
+		sequentialReadSizeMb: sequentialReadSizeMb,
+		rangeReader:          NewRangeReader(obj, bucket, metricHandle),
+		mrr:                  NewMultiRangeReader(obj, metricHandle, mrdWrapper),
+		readType:             util.Sequential,
+	}
+}
+
+func (gr *GCSReader) ReadAt(ctx context.Context, p []byte, offset int64) (gcsx.ReaderResponse, error) {
+	if gr.rangeReader.invalidateReaderIfMisalignedOrTooSmall(offset, p) {
+		gr.seeks++
+	}
+
+	var readerResponse gcsx.ReaderResponse
+	readReq := &gcsx.GCSReaderRequest{
+		Buffer:    p,
+		Offset:    offset,
+		EndOffset: -1,
+	}
+	var err error
+
+	readerResponse, err = gr.rangeReader.readFromExistingReader(ctx, readReq)
+	if err == nil {
+		return readerResponse, nil
+	}
+	if !errors.Is(err, gcsx.FallbackToAnotherReader) {
+		return readerResponse, err
+	}
+
+	// If we don't have a reader, determine whether to read from NewReader or Mrr.
+	end, err := gr.getReadInfo(offset, int64(len(p)))
+	if err != nil {
+		err = fmt.Errorf("ReadAt: getReaderInfo: %w", err)
+		return readerResponse, err
+	}
+
+	readReq.EndOffset = end
+	readerType := gr.getReaderType(offset, end, gr.bucket.BucketType())
+	if readerType == RangeReaderType {
+		readerResponse, err = gr.rangeReader.ReadAt(ctx, readReq)
+		gr.totalReadBytes += uint64(readerResponse.Size)
+		return readerResponse, err
+	}
+
+	readerResponse, err = gr.mrr.ReadAt(ctx, readReq)
+	gr.totalReadBytes += uint64(readerResponse.Size)
+	if err != nil {
+		return readerResponse, err
+	}
+
+	return readerResponse, nil
+}
+
+// readerType specifies the go-sdk interface to use for reads.
```

Comment:
Can we keep the method name same in the comment?

We can get rid of get, as the function returns readerType. Ref: https://google.github.io/styleguide/go/best-practices.html#naming-conventions